### PR TITLE
librbd/migration: prune snapshot extents in RawFormat::list_snaps()

### PR DESCRIPTION
**Pull request by idryomov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by idryomov** [comment]
jenkins test api

**Comment by idryomov** [comment]
Before:

https://pulpito.ceph.com/dis-2024-09-01_18:38:24-rbd-main-distro-default-smithi/

```
Image migration: 99% complete...2024-09-01T19:19:54.276+0000 7fa4effff640 -1 librbd::migration::FileStream::ReadRequest 0x7fa4dc0647a0 read: failed to read 4194304 bytes from file stream
2024-09-01T19:19:54.277 INFO:tasks.workunit.client.0.smithi099.stderr:2024-09-01T19:19:54.276+0000 7fa4effff640 -1 librbd::deep_copy::ObjectCopyRequest: 0x7fa4dc016890 handle_read: failed to read from source object: (34) Numerical result out of range
2024-09-01T19:19:54.277 INFO:tasks.workunit.client.0.smithi099.stderr:2024-09-01T19:19:54.276+0000 7fa4effff640 -1 librbd::io::CopyupRequest: 0x7fa4d0017e50 handle_deep_copy: rbd_data.2.114bab1a11d9.0000000000000100 error encountered during deep-copy: (34) Numerical result out of range
2024-09-01T19:19:54.277 INFO:tasks.workunit.client.0.smithi099.stderr:2024-09-01T19:19:54.276+0000 7fa4effff640 -1 librbd::io::ObjectRequest: 0x7fa4f000a300 handle_copyup: rbd_data.2.114bab1a11d9.0000000000000100 failed to copyup object: (34) Numerical result out of range
2024-09-01T19:19:54.288 INFO:tasks.workunit.client.0.smithi099.stderr:2024-09-01T19:19:54.288+0000 7fa4effff640 -1 librbd::MigrateRequest: 0x7fa4dc05ccf0 handle_migrate_objects: failed to migrate objects: (34) Numerical result out of range
2024-09-01T19:19:54.288 INFO:tasks.workunit.client.0.smithi099.stderr:2024-09-01T19:19:54.288+0000 7fa4effff640 -1 librbd::MigrateRequest: 0x7fa4dc05ccf0 should_complete: encountered error: (34) Numerical result out of range
2024-09-01T19:19:54.288 INFO:tasks.workunit.client.0.smithi099.stderr:2024-09-01T19:19:54.288+0000 7fa4f70caf00 -1 librbd::Migration: execute: migration failed: (34) Numerical result out of range
2024-09-01T19:19:54.292 INFO:tasks.workunit.client.0.smithi099.stderr:
Image migration: 99% complete...failed.
2024-09-01T19:19:54.292 INFO:tasks.workunit.client.0.smithi099.stderr:rbd: migration failed: (34) Numerical result out of range
```

After:

https://pulpito.ceph.com/dis-2024-09-01_18:39:45-rbd-wip-dis-testing-distro-default-smithi/

File: qa/workunits/rbd/cli_migration.sh [hunk 1]
```diff
@@ -27,7 +27,10 @@ expect_false() {
 create_base_image() {
     local image=$1
 
-    rbd create --size 1G ${image}
+    # size is not a multiple of object size to trigger an edge case in
+    # list-snaps
```
Comment (ajarr) [inline qa/workunits/rbd/cli_migration.sh:31]: Is this the edge case that you refer to here https://tracker.ceph.com/issues/67845#note-1 ? 
"This is because deep_copy::ObjectCopyRequest submits list-snaps for 0~m_dst_image_ctx->layout.object_size range." I'm still unable to understand the cause of this edge case that you mention in the tracker comment. Can you dumb it down further?

**Review by ajarr — APPROVED** [review]

File: qa/workunits/rbd/cli_migration.sh (hunk 1 above)
Comment (idryomov) [reply qa/workunits/rbd/cli_migration.sh:31]: > Is this the edge case that you refer to here https://tracker.ceph.com/issues/67845#note-1 ? "This is because deep_copy::ObjectCopyRequest submits list-snaps for 0~m_dst_image_ctx->layout.object_size range."

Yes.  The edge case is an out of bounds read getting submitted to a stream, `ObjectCopyRequest` triggers it.

> I'm still unable to understand the cause of this edge case that you mention in the tracker comment. Can you dumb it down further?

Consider an image whose size is not a multiple of object size, e.g. 21M with standard 4M objects.  `ObjectCopyRequest` always submits list-snaps for entire objects, so when it gets to the last object, it will map 0..4M object extent to 20M..24M image extent and submit list-snaps for that.  If the image is being live imported from a raw format, `RawFormat::list_snaps()` will call `RawSnapshot::list_snap()` on the same extent and (with the current code -- this is getting changed in https://github.com/ceph/ceph/pull/44640) `SPARSE_EXTENT_STATE_DATA` will be returned for 20M..24M range.  `ObjectCopyRequest` will interpret that as an instruction to submit a read for 20M..24M range, generating an out-of-bounds read.

Does this help?

Comment (ajarr) [reply qa/workunits/rbd/cli_migration.sh:31]: Thanks for the explanation.  I understand it better now.
